[CORE-12505] kafka: update list_offsets to return the leader epoch of the offest

IoannisRP opened the pull request:
Fixes: [CORE-12505](https://redpandadata.atlassian.net/browse/CORE-12505)

Changes in this PR:
- expose "offset to leader epoch'" functionality in partition proxy.
- `list_offsets` request now returns the leader epoch of the offset, instead of the partition's. This is for agreement with reference implementation.
- Extend `KCL.list_offsets` function to optionally return leader epochs.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [ ] v24.2.x

## Release Notes

### Bug Fixes

* Fixed a bug where ListOffsets requests would return the current leader epoch, instead of the leader epoch of the offset.

[CORE-12505]: https://redpandadata.atlassian.net/browse/CORE-12505?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
kafka: update list_offsets to return the leader epoch of the offest

list_offsets request now returns the leader epoch of the offset,
instead of the partition's. This is for agreement with reference
implementation.

michael-redpanda commented on tests/rptest/tests/offset_for_leader_epoch_test.py:
```diff
@@ -258,3 +258,58 @@ def all_offsets_valid():
             # Check partition_offsets is a subset of fetched_offsets
             assert fetched_offsets == fetched_offsets | partition_offsets, f"Mismatched offsets for leader epoch {epoch}"
+
+    @skip_debug_mode
```
question: why skip debug mode for this test?

IoannisRP replied:
i was copying the test from above and i forgot it :sweat_smile: 
It is not needed in this test. Removed :+1:

michael-redpanda reviewed (dismissed):
Must change: Please add a bug fix report into the release notes

nit: I think you can break this up into two commits: one for changes to replicated_partition and partition_proxy and another to update the handler.

dotnwat reviewed (commented):
@bharathv did you also mention an issue with epochs in the fetch handler?

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--87a3094c-3464-4d8d-be0a-7f555aff40db-->
<!--TEST RESULTS-->
<!--d415c588-0851-4c4c-9fbc-70b563a8500e-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67906">build#67906</a></summary>


| test_class                  | test_method                             | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:----------------------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest   | test_read_chunks                        |                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67906#0197a879-12fb-44fd-b571-1697707c15d5 | FLAKY         | 19/21    | upstream reliability is '98.81656804733728'. current run reliability is '90.47619047619048'. drift is 8.34038 and the allowed drift is set to 50. The test should PASS |
| ShadowIndexingWhileBusyTest | test_create_or_delete_topics_while_busy | {"cloud_storage_type": 2, "short_retention": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67906#0197a873-f48f-47c1-8ca9-bbccc0fa662a | FLAKY         | 9/12     |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67969">build#67969</a></summary>


| test_class                  | test_method                              | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:-----------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerOffsetsRecoveryTest | test_consumer_offsets_partition_recovery |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67969#0197ac15-e5ec-4f02-8a0f-5e1d2a1ee240 | FLAKY         | 20/21    | upstream reliability is '98.69451697127938'. current run reliability is '95.23809523809523'. drift is 3.45642 and the allowed drift is set to 50. The test should PASS |
| DataMigrationsApiTest       | test_creating_and_listing_migrations     |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67969#0197ac15-e5e9-4491-a263-4073312f66e1 | FLAKY         | 20/21    | upstream reliability is '99.2084432717678'. current run reliability is '95.23809523809523'. drift is 3.97035 and the allowed drift is set to 50. The test should PASS  |
</details>
<!--/body-->

bharathv commented:
> @bharathv did you also mention an issue with epochs in the fetch handler?

Thats @nvartolomei

IoannisRP committed:
kafka: expose offset to leader epoch functionality in partition proxy

IoannisRP committed:
rptests/clients: extend KCL.list_offsets wrapper

Now there is the option to include leader epochs in the reply

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/fc4da5f57cf216f943c7ca2a56e0bca85c32effc..6bd81abde967cbe5390094c857a580525972ce53):
- Removed `skip_debug_mode` for new rptest
- Split commit history to functional and non-functional changes.

pgellert commented on tests/rptest/tests/offset_for_leader_epoch_test.py:
```diff
@@ -258,3 +258,54 @@ def all_offsets_valid():
+            rpk.consume(topic.name, group=group, n=n_records, format=my_format)
+
+        consume_all()
+
```
I think this part is not necessary for what the test is testing, since the epochs exist independently of any consumers
```suggestion
```

IoannisRP replied:
The consumer was acting as the waiting period for everything to fall in place.
Replaced with an actual `wait_until`.

pgellert reviewed (commented):
Other than 1 minor change to the test, this looks good to me

pgellert commented:
@nvartolomei do you have any more context on why it was believed that the fetch implementation was incorrect?
> e.g. in Fetch we validate current leader epoch but now that seems wrong maybe we should validate the epoch for the requested offset

Our implementation is here: https://github.com/redpanda-data/redpanda/blob/e332c36e32c8c526cb8485209d3c967718bbcea2/src/v/kafka/server/handlers/details/leader_epoch.h#L20-L37

and it matches both:
1. what is described in [KIP-320](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=87295403#KIP320:Allowfetcherstodetectandhandlelogtruncation-LeaderFetchHandling):
> The fetch will only be permitted if the requested epoch matches the leader's epoch. If the requested epoch is older than the leader's, we will use a new FENCED_LEADER_EPOCH error code. If the epoch is newer than the leader's (for example in a [KIP-232](https://cwiki.apache.org/confluence/display/KAFKA/KIP-232%3A+Detect+outdated+metadata+using+leaderEpoch+and+partitionEpoch) scenario), we will use a new UNKNOWN_LEADER_EPOCH error code.

2. as well as the apache kafka implementation here: https://github.com/apache/kafka/blob/56aeaa4c44c5c3ddcd169b1c5188e161b08f3c14/core/src/main/scala/kafka/cluster/Partition.scala#L514-L527

michael-redpanda reviewed (commented):
lgtm -

once the initial CI is completed, please run the new test a bunch of times using `/ci-repeat` to ensure we aren't introducing any flakiness.  Thanks!

IoannisRP commented:
/ci-repeat 10
tests/rptest/tests/offset_for_leader_epoch_test.py::OffsetForLeaderEpochTest.test_leader_epoch_for_offset

nvartolomei commented on src/v/kafka/data/replicated_partition.cc:
```diff
@@ -431,6 +431,11 @@ replicated_partition::get_leader_epoch_last_offset(
 
+kafka::leader_epoch
+replicated_partition::get_leader_epoch(model::offset o) const {
+    return leader_epoch_from_term(_partition->get_term(o));
```
Missing cloud/tiered storage implementation? As well as test.

nvartolomei commented:
@pgellert I don't remember the context. Something just seemed off to me at the time. Btw, here is how kafka does validate the epochs in the request https://github.com/apache/kafka/blob/cff10e654149ed2d4924235a15156d4bc0bf7b4a/core/src/main/scala/kafka/cluster/Partition.scala#L1520

Note that there is a currentLeaderEpoch and a lastFetchedEpoch. Currently there are no references to lastFetchedEpoch field in our codebase (generated at `bazel-out/k8-fastbuild/bin/src/v/kafka/protocol/schemata/_virtual_includes/fetch_request/kafka/protocol/schemata/fetch_request.h#L45`) from json definition https://github.com/redpanda-data/redpanda/blob/a2ae70e6f4e3b47cd11edd9c65fbfd9576873004/src/v/kafka/protocol/schemata/fetch_request.json#L84.

Just highlighting. I don't know if this is relevant in our implementation vs kafka where the fetch protocol is also used for replication afaik.

pgellert commented:
@nvartolomei Thanks. We don't yet support the `LastFetchedEpoch` field and the associated validation, since we're at fetch [v11](https://github.com/redpanda-data/redpanda/blob/6e484eaa743378e6ffc7a8b0f3781177697f497e/src/v/kafka/server/handlers/fetch.h#L36) and that is introduced in [v12](https://github.com/redpanda-data/redpanda/blob/a2ae70e6f4e3b47cd11edd9c65fbfd9576873004/src/v/kafka/protocol/schemata/fetch_request.json#L47). [KIP-595](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=148647470#KIP595:ARaftProtocolfortheMetadataQuorum-FetchRequestHandling) introduced fetch v12 and the [pre-KIP-595 logic](https://github.com/apache/kafka/blob/71e85f5e842687c040f9ee69b9073fc285113541/core/src/main/scala/kafka/cluster/Partition.scala#L770) lines up with what we have, so I think we're good here.

tyson-redpanda commented:
Going with a different approach.
